### PR TITLE
Handling Operation Policy Spec and Definition

### PR DESCRIPTION
**Pull request by renuka-fernando**
### Purpose
$subject

Sample Spec

```yaml
type: operation_policy_specification
version: v4.1.0
data:
  category: Mediation
  name: addRequestHeader
  displayName: Set Header
  description: With this policy, user can add a new header to the request
  applicableFlows:
    - request
  supportedGateways:
    - CC
  supportedApiTypes:
    - REST
  policyAttributes:
    -
    name: headerName
    displayName: Header Name
    description: Name of the header to be added
    validationRegex: /^\S+$/
    type: String
    required: true
    -
    name: headerValue
    displayName: Header Value
    description: Value of the header
    validationRegex: ""
    type: String
    required: true
  multipleAllowed: true
```

Sample Policy
```yaml
definition:
  action: SET_HEADER
  parameters:
    headerName: {{ .headerName }}
    headerValue: {{ .headerValue }}
```

```yaml
definition:
  action: CALL_INTERCEPTOR_SERVICE
  parameters:
    interceptorServiceURL: {{ .interceptorServiceURL }}
    includes:
    {{- range $i, $val := .includes  }}
      - {{ $val }}
    {{- end }}
```

### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2655

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: No

### Tested environments

Docker Compose

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/oasparser/model/policy_container.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	t, err := template.New("policy-def").Parse(string(defRaw))
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error parsing the operation policy definition \"%v\" into go template: %v", policy.PolicyName, err)
+		return Policy{}, err
+	}
+
+	var out bytes.Buffer
+	err = t.Execute(&out, policy.Parameters)
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error operation policy definition \"%v\": %v", policy.PolicyName, err)
```
Comment (VajiraPrabuddhaka) [inline adapter/internal/oasparser/model/policy_container.go:110]: Use the formalized error log format..

File: adapter/internal/api/apis_process.go [hunk 2]
```diff
@@ -128,6 +131,29 @@ func processFileInsideProject(apiProject *model.ProjectAPI, fileContent []byte,
...
+		policy := model.PolicyContainer{}
+		if _, ok := apiProject.Policies[policyName]; ok {
+			policy = apiProject.Policies[policyName]
+		}
+
+		if strings.HasSuffix(fileName, policySpecFileExtension) {
+			// process policy specificationn
+			spec := model.PolicySpecification{}
+			if err := yaml.Unmarshal(fileContent, &spec); err != nil { // check that, JSON files are also supported
+				loggers.LoggerAPI.Errorf("Error parsing content of policy specification %v: %v", fileName, err.Error())
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/api/apis_process.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/febace122f95f303868e309a6bbabfd6340c8a35

File: adapter/internal/oasparser/model/policy_container.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	spec := p[policy.PolicyName].Specification
+	if err := spec.validatePolicy(policy, flow, stats, index); err != nil {
+		loggers.LoggerAPI.Errorf("Operation policy validation failed, ignoring the policy \"%v\": %v", policy.PolicyName, err)
+		return policy, err
+	}
+
+	defRaw := p[policy.PolicyName].Definition.RawData
+	t, err := template.New("policy-def").Parse(string(defRaw))
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error parsing the operation policy definition \"%v\" into go template: %v", policy.PolicyName, err)
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/febace122f95f303868e309a6bbabfd6340c8a35

File: adapter/internal/oasparser/model/policy_container.go (hunk 1 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/febace122f95f303868e309a6bbabfd6340c8a35

File: adapter/internal/oasparser/model/policy_container.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	var out bytes.Buffer
+	err = t.Execute(&out, policy.Parameters)
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error operation policy definition \"%v\": %v", policy.PolicyName, err)
+		return Policy{}, err
+	}
+
+	def := PolicyDefinition{}
+	if err := yaml.Unmarshal(out.Bytes(), &def); err != nil {
+		loggers.LoggerAPI.Errorf("Error parsing standardized operation policy definition \"%v\" into yaml: %v", policy.PolicyName, err)
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/febace122f95f303868e309a6bbabfd6340c8a35

File: adapter/internal/oasparser/model/policy_container.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+	return fmtPolicies
+}
+
+// getFormattedPolicyFromTemplated returns formatted, Choreo Connect policy from a user templated policy
+func (p PolicyContainerMap) getFormattedPolicyFromTemplated(policy Policy, flow PolicyFlow, stats map[string]policyStats, index int) (Policy, error) {
+	// using index i instead of struct to validate policy against multiple allowed and apply only first if multiple exists
+	spec := p[policy.PolicyName].Specification
+	if err := spec.validatePolicy(policy, flow, stats, index); err != nil {
+		loggers.LoggerAPI.Errorf("Operation policy validation failed, ignoring the policy \"%v\": %v", policy.PolicyName, err)
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/febace122f95f303868e309a6bbabfd6340c8a35

File: adapter/internal/oasparser/model/policy_container.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+			fmtPolicies.Fault = append(fmtPolicies.In, fmtPolicy)
+		}
+	}
+
+	return fmtPolicies
+}
+
+// getFormattedPolicyFromTemplated returns formatted, Choreo Connect policy from a user templated policy
+func (p PolicyContainerMap) getFormattedPolicyFromTemplated(policy Policy, flow PolicyFlow, stats map[string]policyStats, index int) (Policy, error) {
+	// using index i instead of struct to validate policy against multiple allowed and apply only first if multiple exists
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go:94]: Here we use `policyStats` to validate policies, when `multiple allowed` is disabled and the user provides multiple policies of that spec. Yes, we can just ignore this validation but feels better to log validation errors IMO.

File: adapter/internal/api/apis_process.go [hunk 7]
```diff
@@ -128,6 +131,29 @@ func processFileInsideProject(apiProject *model.ProjectAPI, fileContent []byte,
...
+		policyName := utills.FileNameWithoutExtension(fileName)
+		policy := model.PolicyContainer{}
+		if _, ok := apiProject.Policies[policyName]; ok {
+			policy = apiProject.Policies[policyName]
+		}
+
+		if strings.HasSuffix(fileName, policySpecFileExtension) {
+			// process policy specificationn
+			spec := model.PolicySpecification{}
+			if err := yaml.Unmarshal(fileContent, &spec); err != nil { // check that, JSON files are also supported
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/api/apis_process.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/b9a42beb540356265589d6acf432571632dd194c

File: adapter/internal/oasparser/model/policy_container.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+	policy.Action = def.Definition.Action
+
+	return policy, nil
+}
+
+// PolicySpecification holds policy specification from ./Policy/<policy>.yaml files
+type PolicySpecification struct {
+	Type    string   `yaml:"type" json:"type"`
+	Version string   `yaml:"version" json:"version"`
+	Data    struct { // TODO: check all fields are required or omit empty in APIM side
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/b9a42beb540356265589d6acf432571632dd194c

File: adapter/internal/api/apis_impl.go [hunk 9]
```diff
@@ -56,6 +56,7 @@ func extractAPIProject(payload []byte) (apiProject model.ProjectAPI, err error)
 	// TODO: (VirajSalaka) this won't support for distributed openAPI definition
 	apiProject.UpstreamCerts = make(map[string][]byte)
 	apiProject.EndpointCerts = make(map[string]string)
+	apiProject.Policies = make(map[string]model.PolicyContainer)
```
Comment (suksw) [inline adapter/internal/api/apis_impl.go:59]: Let's add this here too https://github.com/wso2/product-microgateway/blob/348cad1cd2dbedf32672234c4cdb1990462dfd3e/adapter/internal/api/apis_impl.go#L95

File: adapter/internal/oasparser/model/api_yaml.go [hunk 10]
```diff
@@ -97,17 +97,39 @@ type OperationYaml struct {
...
+// policyStats used to optimize and reduce loops by storing stats by calculating only once
+type policyStats struct {
+	firstIndex int
+	count      int
+}
+
+// PolicyList holds list of Polices in a flow of operation
+type PolicyList []Policy
+
+func (pl PolicyList) getStats() map[string]policyStats {
```
Comment (suksw) [inline adapter/internal/oasparser/model/api_yaml.go:114]: Shall we move this file after the constructor method for the apiyaml, and after the necessary methods for a regular API deployment. Let's go priority wise.

File: adapter/internal/oasparser/model/policy_container.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+	for i, policy := range policies.In {
+		if fmtPolicy, err := p.getFormattedPolicyFromTemplated(policy, policyInFlow, inFlowStats, i); err == nil {
+			fmtPolicies.In = append(fmtPolicies.In, fmtPolicy)
+		}
+	}
+
+	outFlowStats := policies.Out.getStats()
+	for i, policy := range policies.Out {
+		if fmtPolicy, err := p.getFormattedPolicyFromTemplated(policy, policyOutFlow, outFlowStats, i); err == nil {
+			fmtPolicies.Out = append(fmtPolicies.In, fmtPolicy)
```
Comment (suksw) [inline adapter/internal/oasparser/model/policy_container.go:78]:
```suggestion
			fmtPolicies.Out = append(fmtPolicies.Out, fmtPolicy)
```
should this change?

File: adapter/internal/oasparser/model/policy_container.go [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+		}
+	}
+
+	faultFlowStats := policies.Out.getStats()
```
Comment (suksw) [inline adapter/internal/oasparser/model/policy_container.go:82]:
```suggestion
	faultFlowStats := policies.Fault.getStats()
```
Should this change?

File: adapter/internal/oasparser/model/policy_container.go [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+	for i, policy := range policies.Fault {
+		if fmtPolicy, err := p.getFormattedPolicyFromTemplated(policy, policyFaultFlow, faultFlowStats, i); err == nil {
+			fmtPolicies.Fault = append(fmtPolicies.In, fmtPolicy)
```
Comment (suksw) [inline adapter/internal/oasparser/model/policy_container.go:85]:
```suggestion
			fmtPolicies.Fault = append(fmtPolicies.Fault, fmtPolicy)
```
Should this change?

File: adapter/internal/oasparser/model/policy_container.go [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+		Action     string                 `yaml:"action"`
+		Parameters map[string]interface{} `yaml:"parameters"`
+	} `yaml:"definition"`
+	RawData []byte `yaml:"-"`
+}
+
+// validatePolicy validates the given policy against the spec
+func (spec *PolicySpecification) validatePolicy(policy Policy, flow PolicyFlow, stats map[string]policyStats, index int) error {
+	if spec.Data.Name != policy.PolicyName {
+		return fmt.Errorf("invalid policy specification, spec name %s and policy name %s mismatch", spec.Data.Name, policy.PolicyName)
```
Comment (pubudu538) [inline adapter/internal/oasparser/model/policy_container.go:175]: Can we improve these loggings? We need to identify the API, org, etc.  We can use the new logging methods.

File: adapter/internal/oasparser/model/policy_container.go [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+// PolicyDefinition holds the content of policy definition which is rendered from ./Policy/<policy>.gotmpl files
+type PolicyDefinition struct {
+	Definition struct {
+		Action     string                 `yaml:"action"`
+		Parameters map[string]interface{} `yaml:"parameters"`
+	} `yaml:"definition"`
+	RawData []byte `yaml:"-"`
+}
```
Comment (suksw) [inline adapter/internal/oasparser/model/policy_container.go:170]: Shall we move these to the top. Maybe after constants. So then when people don't want to know any of the impl details, they can just read these as soon as they open the file.

File: adapter/internal/oasparser/model/policy_container.go [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	// TODO: check ApplicableFlows are (in, out, fault) or (request, response, fault)
+	if !arrayContains(spec.Data.ApplicableFlows, string(flow)) {
+		return fmt.Errorf("policy flow \"%s\" not supported", flow)
+	}
+	if !arrayContains(spec.Data.SupportedGateways, policyCCGateway) {
+		return errors.New("choreo connect gateway not supported")
+	}
+	if !spec.Data.MultipleAllowed {
+		// TODO: check the behaviour with APIM
```
Comment (suksw) [inline adapter/internal/oasparser/model/policy_container.go:185]:
```suggestion
		// TODO (renuka): check the behaviour with APIM
```

File: adapter/internal/oasparser/model/policy_container.go (hunk 15 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/321b5c37e580d65072fbcd8902e950a87217755a

File: adapter/internal/oasparser/model/policy_container.go (hunk 13 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/321b5c37e580d65072fbcd8902e950a87217755a

File: adapter/internal/oasparser/model/policy_container.go (hunk 11 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/321b5c37e580d65072fbcd8902e950a87217755a

File: adapter/internal/oasparser/model/api_yaml.go (hunk 10 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/api_yaml.go]: fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/321b5c37e580d65072fbcd8902e950a87217755a

File: adapter/internal/oasparser/model/policy_container.go (hunk 14 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/policy_container.go]: added with https://github.com/wso2/product-microgateway/pull/2656/commits/cb46ca154bdeb62c3ce0e3a3a24c0295f9aedcea

File: adapter/internal/oasparser/model/policy_container.go [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+
+const (
+	policyCCGateway string = "CC"
+)
+
+// policy value validation types
+const (
+	policyValTypeString string = "String"
+	policyValTypeInt    string = "Integer"
+	policyValTypeBool   string = "Boolean" // TODO: check type names with APIM
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/policy_container.go:41]: please follow todo(name) comment format

File: adapter/internal/oasparser/model/api_yaml.go [hunk 18]
```diff
@@ -97,17 +97,26 @@ type OperationYaml struct {
...
+type PolicyList []Policy
+
 // Policy holds APIM policies
 type Policy struct {
-	PolicyName   string      `json:"policyName,omitempty"`
-	TemplateName string      `json:"templateName,omitempty"`
-	Order        int         `json:"order,omitempty"`
-	Parameters   interface{} `json:"parameters,omitempty"`
+	PolicyName string      `json:"policyName,omitempty"`
+	Action     string      `json:"-,omitempty"`
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/api_yaml.go:117]:
```suggestion
	Action     string     
```
what if we remove this json / yaml tag

File: adapter/internal/oasparser/model/policy_container.go [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+	}
+}
+
+// PolicyDefinition holds the content of policy definition which is rendered from ./Policy/<policy>.gotmpl files
+type PolicyDefinition struct {
+	Definition struct {
+		Action     string                 `yaml:"action"`
+		Parameters map[string]interface{} `yaml:"parameters"`
+	} `yaml:"definition"`
+	RawData []byte `yaml:"-"`
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/policy_container.go:89]:
```suggestion
	RawData []byte 
```

File: adapter/internal/oasparser/model/policy_container.go [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+		for _, attrib := range spec.Data.PolicyAttributes {
+			val, found := policyPrams[attrib.Name]
+			if !found {
+				if attrib.Required {
+					return fmt.Errorf("required paramater %s not found", attrib.Name)
+				}
+				continue
+			}
+
+			// TODO: check this Value and Regex validation is needed
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/policy_container.go:208]: here as well

File: adapter/internal/oasparser/model/policy_container.go [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+				if !strings.EqualFold(attrib.Type, policyValTypeString) {
+					return fmt.Errorf("invalid value type of paramater %s, required %s", attrib.Name, attrib.Type)
+				}
+				regexStr := attrib.ValidationRegex
+				if regexStr != "" {
+					if !strings.HasPrefix(regexStr, "/") || !strings.HasSuffix(regexStr, "/") {
+						return fmt.Errorf("invalid regex expression in policy spec %s, regex: \"%s\", regex expression should starts and end with '/'", spec.Data.Name, attrib.ValidationRegex)
+					}
+					regexStr = regexStr[1 : len(regexStr)-1]
+
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/policy_container.go:220]:
```suggestion
```

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Review by suksw — APPROVED** [review]

File: adapter/internal/oasparser/model/api_yaml.go (hunk 18 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/api_yaml.go:117]: when we use the name "-" yaml/json parser do not marshal/unmarshal it. we can remove `omitempty` here

**Pull request merged**